### PR TITLE
async fetch hook

### PR DESCRIPTION
**Pull request by xinhaoz**
Hook for async data fetching, written in typescript

### CODE REVIEW

File: src/Components/Admin/UserManagement.jsx [hunk 1]
```diff
@@ -103,27 +85,4 @@ function UserManagement() {
...
-//       let programs = [];
-//       programs.push(p);
-//       let user = {
-//         userId: item.uid,
-//         name: "",
-//         email: item.email,
-//         role: "User",
-//         programs: programs
-//       };
-//       //2 FACTOR AUTH. CONTACT GREG BEFORE UNCOMMENTING
```
Comment (GodGreg) [inline src/Components/Admin/UserManagement.jsx:123]: Thanks for contacting me :). this if fine tho

File: src/Hooks/useAsyncFetch.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      return { ...state, error: action.error, isPending: false };
+    default:
+      return state;
+  }
+}
+
+export default function useAsyncFetch<ResultType>(
+  queryFunc: (params?: any) => Promise<ResultType>,
+  params?: object | null,
+  defaultValue?: ResultType
```
Comment (GodGreg) [inline src/Hooks/useAsyncFetch.ts:50]: Could you explain what defaultValue is/does? Having a hard time following along :P

File: src/Components/Admin/UserManagement.jsx [hunk 3]
```diff
@@ -32,56 +31,39 @@ const Header = styled.div`
...
 }
 
 function UserManagement() {
-  const [users, setUsers] = useState(null);
-
-  useEffect(() => {
-    async function getUsers() {
-      // Fetch the users from the backend
-      const fetched = await GET.getAllUsersAPI();
+  const [loadUsers] = useAsyncFetch(GET.getAllUsersAPI, {}, []);
```
[the remark this answers is not in the document]
Comment (GodGreg) [reply src/Components/Admin/UserManagement.jsx:57]: I guess you haven't seen my changes yet for the API calls. I created a helper.js file to modularize common code. It is on the branch greg/scaling. 

https://github.com/uwblueprint/cookie-monster/blob/17c77ecc72df98220d8b07b11fef80a533c192fa/src/requests/Helper.js

File: src/Hooks/useAsyncFetch.ts (hunk 2 above)
Comment (xinhaoz) [reply src/Hooks/useAsyncFetch.ts:13]: It's what value is initially set to, similar to useState(defaultValue)

File: src/Components/AllCandidates/CommitteeReview.jsx [hunk 4]
```diff
@@ -8,7 +8,13 @@ import { faAngleLeft } from "@fortawesome/free-solid-svg-icons";
 import { connect } from "react-redux";
 import Spinner from "react-spinner-material";
 
-const GET = require("../../requests/get");
+import useAsyncFetch from "../../Hooks/useAsyncFetch";
+
+import {
+  getAllUsersAPI,
```
[the remark this answers is not in the document]
Comment (xinhaoz) [reply src/Components/AllCandidates/CommitteeReview.jsx:14]: I'm more inclined to explicitly import what we need from the module, feel like it's easier for refactoring. Mostly just personal preference though.

File: src/Components/Admin/UserManagement.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (xinhaoz) [reply src/Components/Admin/UserManagement.jsx]: I took a look at your branch, is the goal to later just use GET(url) from a component?

[the remark this answers is not in the document]
Comment (GodGreg) [reply src/Components/Admin/UserManagement.jsx]: I love the elegance of this code. 

I am hesitant to have URLs defined within components (especially if we use it in multiple places). I personally think it is better to define each call in get.js for maintainability even if each function is just calling GET from helper.js. 
```
async function getAllRankingsAPI() {
  return GET(`/api/stackings`)
}
```

Also having two separate ways to do an API call like this is probably confusing to other devs. I think I'd prefer to stick with just usePromise.

[the remark this answers is not in the document]
Comment (xinhaoz) [reply src/Components/Admin/UserManagement.jsx]: Okay, SGTM!

**Pull request merged**